### ISSUE TITLE
[BUG] conflict with torchlib

### ISSUE THREAD

**Issue by pommedeterresautee** [NVIDIA/cutlass#578]
**Describe the bug**

I try to modify the first example (01) to test PyTorch tensors as inputs.
When I add libtorch dependency in `cmakelists` I get an error message making me thinking there is a conflict of version or some variable override related to CUDA version to use.

**Steps/Code to reproduce bug**

```cmake
# cmakelists from 00 basic gemm

cutlass_example_add_executable(
  00_basic_gemm
  basic_gemm.cu
)

set(CMAKE_PREFIX_PATH "/hardcoded/absolute/path/to/libtorch")
set(Torch_DIR "/hardcoded/absolute/path/to/libtorch")
find_package(Torch REQUIRED)
target_link_libraries(00_basic_gemm "${TORCH_LIBRARIES}")
```

produces:

```
[main] Configuring folder: cutlass 
[proc] Executing command: /usr/bin/cmake --no-warn-unused-cli -DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL=TRUE -DCMAKE_BUILD_TYPE:STRING=Debug -DCMAKE_C_COMPILER:FILEPATH=/usr/bin/gcc -DCMAKE_CXX_COMPILER:FILEPATH=/usr/bin/g++ -S/home/geantvert/workspace/cutlass -B/home/geantvert/workspace/cutlass/build -G "Unix Makefiles"
[cmake] Not searching for unused variables given on the command line.
[cmake] -- CMake Version: 3.22.1
[cmake] -- CUDART: /usr/local/cuda-11/lib64/libcudart.so
[cmake] -- CUDA Driver: /usr/local/cuda-11/lib64/stubs/libcuda.so
[cmake] -- NVRTC: /usr/local/cuda-11/lib64/libnvrtc.so
[cmake] -- Default Install Location: install
[cmake] CMake Deprecation Warning at build/_deps/googletest-src/CMakeLists.txt:1 (cmake_minimum_required):
[cmake]   Compatibility with CMake < 2.8.12 will be removed from a future version of
[cmake]   CMake.
[cmake] 
[cmake]   Update the VERSION argument <min> value or use a ...<max> suffix to tell
[cmake]   CMake that the project does not need compatibility with older versions.
[cmake] 
[cmake] 
[cmake] CMake Deprecation Warning at build/_deps/googletest-src/googlemock/CMakeLists.txt:46 (cmake_minimum_required):
[cmake]   Compatibility with CMake < 2.8.12 will be removed from a future version of
[cmake]   CMake.
[cmake] 
[cmake]   Update the VERSION argument <min> value or use a ...<max> suffix to tell
[cmake]   CMake that the project does not need compatibility with older versions.
[cmake] 
[cmake] 
[cmake] CMake Deprecation Warning at build/_deps/googletest-src/googletest/CMakeLists.txt:55 (cmake_minimum_required):
[cmake]   Compatibility with CMake < 2.8.12 will be removed from a future version of
[cmake]   CMake.
[cmake] 
[cmake]   Update the VERSION argument <min> value or use a ...<max> suffix to tell
[cmake]   CMake that the project does not need compatibility with older versions.
[cmake] 
[cmake] 
[cmake] -- CUDA Compilation Architectures: 53;60;61;70;72;75;80;86
[cmake] -- Enable caching of reference results in conv unit tests
[cmake] -- Enable rigorous conv problem sizes in conv unit tests
[cmake] -- Using NVCC flags: -DCUTLASS_TEST_LEVEL=0;-DCUTLASS_TEST_ENABLE_CACHED_RESULTS=1;-DCUTLASS_CONV_UNIT_TEST_RIGOROUS_SIZE_ENABLED=1;-DCUTLASS_DEBUG_TRACE_LEVEL=0;$<$<BOOL:1>:-Xcompiler=-Wconversion>;$<$<BOOL:1>:-Xcompiler=-fno-strict-aliasing>;-lineinfo
[cmake] fatal: ni ceci ni aucun de ses répertoires parents (jusqu'au point de montage /) n'est un dépôt git
[cmake] Arrêt à la limite du système de fichiers (GIT_DISCOVERY_ACROSS_FILESYSTEM n'est pas défini).
[cmake] -- CUTLASS Revision: Unable to detect, Git returned code 128.
[cmake] -- Configuring cublas ...
[cmake] -- cuBLAS: /usr/local/cuda-11/lib64/libcublas.so
[cmake] -- cuBLAS: /usr/local/cuda-11/include
[cmake] -- Configuring cuBLAS ... done.
[cmake] -- Configuring cuDNN ...
[cmake] -- cuDNN: /usr/lib/x86_64-linux-gnu/libcudnn.so
[cmake] -- cuDNN: /usr/include
[cmake] -- Configuring cuDNN ... done.
[cmake] -- Caffe2: CUDA detected: 11.7
[cmake] -- Caffe2: CUDA nvcc is: /usr/local/cuda-11/bin/nvcc
[cmake] -- Caffe2: CUDA toolkit directory: /usr/local/cuda-11
[cmake] -- Caffe2: Header version is: 11.7
[cmake] -- Configuring incomplete, errors occurred!
[cmake] See also "/home/geantvert/workspace/cutlass/build/CMakeFiles/CMakeOutput.log".
[cmake] CMake Error at libtorch/share/cmake/Caffe2/public/cuda.cmake:100 (message):
[cmake]   FindCUDA says CUDA version is (usually determined by nvcc), but the CUDA
[cmake]   headers say the version is 11.7.  This often occurs when you set both
[cmake]   CUDA_HOME and CUDA_NVCC_EXECUTABLE to non-standard locations, without also
[cmake]   setting PATH to point to the correct nvcc.  Perhaps, try re-running this
[cmake]   command again with PATH=/usr/local/cuda-11/bin:$PATH.  See above log
[cmake]   messages for more diagnostics, and see
[cmake]   https://github.com/pytorch/pytorch/issues/8092 for more details.
[cmake] Call Stack (most recent call first):
[cmake]   libtorch/share/cmake/Caffe2/Caffe2Config.cmake:88 (include)
[cmake]   libtorch/share/cmake/Torch/TorchConfig.cmake:68 (find_package)
[cmake]   examples/00_basic_gemm/CMakeLists.txt:54 (find_package)
[cmake] 
```

**Expected behavior**
It should compile without crashing.  
The added lines works for other project based on `libtorch`.  

I have of course tried the recommendations in the error message and the PR (running with env var PATH and CUDA_HOME modified, etc.), it changed nothing.

**Environment details (please complete the following information):**
Bare metal, cmake 3.22 on Ubuntu 22.04

**Additional context**
N/A

**Comment by apuaaChen** [comment]
Hi! I find some possible way to diagnose the bug. First, there is some issue with the `libtorch/share/cmake/Caffe2/public/cuda.cmake` at line 91-105 (in my case).  The original code is
```python
if(NOT cuda_version_from_header STREQUAL ${CUDA_VERSION_STRING})
      # Force CUDA to be processed for again next time
      # TODO: I'm not sure if this counts as an implementation detail of
      # FindCUDA
      set(${cuda_version_from_findcuda} ${CUDA_VERSION_STRING})
      unset(CUDA_TOOLKIT_ROOT_DIR_INTERNAL CACHE)
      # Not strictly necessary, but for good luck.
      unset(CUDA_VERSION CACHE)
      # Error out
      message(FATAL_ERROR "FindCUDA says CUDA version is ${cuda_version_from_findcuda} (usually determined by nvcc), "
        "but the CUDA headers say the version is ${cuda_version_from_header}.  This often occurs "
        "when you set both CUDA_HOME and CUDA_NVCC_EXECUTABLE to "
        "non-standard locations, without also setting PATH to point to the correct nvcc.  "
        "Perhaps, try re-running this command again with PATH=${CUDA_TOOLKIT_ROOT_DIR}/bin:$PATH.  "
        "See above log messages for more diagnostics, and see https://github.com/pytorch/pytorch/issues/8092 for more details.")
    endif()
```
However, `set(${cuda_version_from_findcuda} ${CUDA_VERSION_STRING})` doesn't pass the value of ${CUDA_VERSION_STRING} to `cuda_version_from_findcuda`, that's why the version found by FindCUDA is not printed in the error message. The correct version is `set(cuda_version_from_findcuda ${CUDA_VERSION_STRING})`.

Then let's back to the bug. The `cuda_version_from_findcuda` is the exact version of nvcc, e.g. 11.3.123 in my case. On the other hand, `cuda_version_from_header` is printed from a simple script in the `cuda.cmake` file
```python
file(WRITE ${file} ""
    "#include <cuda.h>\n"
    "#include <cstdio>\n"
    "int main() {\n"
    "  printf(\"%d.%d\", CUDA_VERSION / 1000, (CUDA_VERSION / 10) % 100);\n"
    "  return 0;\n"
    "}\n"
    )
```
This will return 11.3 rather than 11.3.123, and then we have cuda_version_from_header != cuda_version_from_findcuda. I found that if I change `"%d.%d` to `"%d.%d.123`, no bug is reported here.

**Comment by pommedeterresautee** [comment]
Nice investigation @apuaaChen !!! Thank you a lot.
Indeed hardcoding the full version numbers avoid the error on my side too.

But it led me in another error (that I have not yet investigated):

```log
-- /usr/local/cuda-11/lib64/libnvrtc.so shorthash is d833c4f3
CMake Error in /home/geantvert/workspace/cutlass/build/CMakeFiles/CMakeTmp/CMakeLists.txt:
  CUDA_ARCHITECTURES is empty for target "cmTC_07b93".


CMake Error in /home/geantvert/workspace/cutlass/build/CMakeFiles/CMakeTmp/CMakeLists.txt:
  CUDA_ARCHITECTURES is empty for target "cmTC_07b93".


CMake Error at libtorch/share/cmake/Caffe2/Modules_CUDA_fix/upstream/FindCUDA/select_compute_arch.cmake:137 (try_run):
  Failed to generate test project build system.
Call Stack (most recent call first):
  libtorch/share/cmake/Caffe2/Modules_CUDA_fix/upstream/FindCUDA/select_compute_arch.cmake:197 (CUDA_DETECT_INSTALLED_GPUS)
  libtorch/share/cmake/Caffe2/public/utils.cmake:401 (cuda_select_nvcc_arch_flags)
  libtorch/share/cmake/Caffe2/public/cuda.cmake:437 (torch_cuda_get_nvcc_gencode_flag)
  libtorch/share/cmake/Caffe2/Caffe2Config.cmake:88 (include)
  libtorch/share/cmake/Torch/TorchConfig.cmake:68 (find_package)
  examples/00_basic_gemm/CMakeLists.txt:38 (find_package)
```

If, as you showed it perfectly, the cmake part of libtorch is not perfect, my main question is why does it work out of the box when it is run outside CUTLASS project.

For instance, the tuto code from libtorch (https://pytorch.org/tutorials/advanced/cpp_frontend.html) run out of the box on my machine at least, it doesn't have version error or unsupported hardware.

Have you been able to run a basic app generating 2 CUDA torch tensors and sending it to cublass for a basic gemm?
Or do you know where I can find such working code?

**Comment by hwu36** [comment]
I think it would be best to ask in the torch community.  We are torch users, not torch developers.  I know many users have success in using cutlass together with torch, you may get help there.

Cutlass has to support a large matrix of different version of {nvcc, nvrtc, msvc, clang, nvc++} x different version of hardware.  Our cmake is one of the best in doing this job.  Not every CUDA library can do what we do.

**Comment by pommedeterresautee** [comment]
Thank you @hwu36 for your response.
Just to be clear, both projects are used in sensitive contexts, and I have no doubt that both `cmake` are well written. My point was just stating that alone each `cmake` works well but, on my machine and my own code, mixing them together led to some issue.

I tried to continue the excellent investigation from @apuaaChen and below are my findings so far (in case it may help someone who is interested in this issue).

Compiling libtorch and cutlass together (which at that point was crashing), the `CUDA_VERSION_STRING` value is `11.7.64` (on my machine + CUDA 11.7).

Compiling libtorch without cutlass and logging `CUDA_VERSION_STRING` value showed `11.7`

So the issue is related to `CUDA_VERSION_STRING` and not `cuda_version_from_header`.  
This variable is instantiated by `FindCuda` command (from `cmake` itself) and is documented here: https://cmake.org/cmake/help/latest/module/FindCUDA.html#result-variables

```
CUDA_VERSION, CUDA_VERSION_STRING
Full version in the X.Y format.
```

The value 11.7.64 is not following the expected format X.Y and is probably to be fixed.

Digging a bit more, it appears that libtorch has its own `FindCuda` implementation not updated since January 2021.
[https://github.com/pytorch/pytorch/blob/master/cmake/Modules_CUDA_fix/upstream/FindCUDA.cmake](https://github.com/pytorch/pytorch/blob/1.6_doc_references/cmake/Modules_CUDA_fix/upstream/FindCUDA.cmake)

This file parses the output of `nvcc —version` and is responsible to set the variable `CUDA_VERSION_STRING`.

When compiled **without** cutlass, variable `CUDA_VERSION` is not set (during the first run only), and on the `If` of line https://github.com/pytorch/pytorch/blob/master/cmake/Modules_CUDA_fix/upstream/FindCUDA.cmake#L784 it goes through the `Then` branch. In this branch, the following command is run:

```
set(CUDA_VERSION "${CUDA_VERSION_MAJOR}.${CUDA_VERSION_MINOR}" CACHE STRING "Version of CUDA as computed from nvcc.")
```

> it sets `CUDA_VERSION` to 11.7

Later on, after the `If`, `CUDA_VERSION_STRING` takes the value of `CUDA_VERSION`

When compiled **with** cutlass, variable `CUDA_VERSION` is set to `11.7.64`, and it goes through the `Else` branch of the same `If`. In this branch,`CUDA_VERSION` is not set to major.minor format.
And that’s how things go bad!

On second run, in libtorch dependency scenario alone, `CUDA_VERSION` is set to 11.7, go `Else` branch but it's ok for obvious reason. In cutlass + libtorch, it's always 11.7.64.
The fix seems to be very simple, as minor and major part of the version are well parsed it should be used to define `CUDA_VERSION_STRING` instead of reusing `CUDA_VERSION`. That way it would match the documented format.

As guessed by @hwu36, there is nothing to do on cutlass side (at most this issue will provide a little documentation for those interested).

I will post findings related to the architecture issue when I have the full story.  
So far, I just know that it works if some values are hardcoded. Again cutlass alone works, so it’s probably related to the interaction with libtorch.

**Comment by pommedeterresautee** [comment]
Regarding the architecture issue, it seems to be a similar issue.
cmake logsafter fixing cuda version as described above:

```log
[cmake] -- Found cuDNN: v8.4.1  (include: /usr/include, library: /usr/lib/x86_64-linux-gnu/libcudnn.so)
[cmake] -- /usr/local/cuda-11/lib64/libnvrtc.so shorthash is d833c4f3
[cmake] CMake Error in /home/geantvert/workspace/cutlass/build/CMakeFiles/CMakeTmp/CMakeLists.txt:
[cmake]   CUDA_ARCHITECTURES is empty for target "cmTC_89c35".
[cmake] 
[cmake] 
[cmake] CMake Error in /home/geantvert/workspace/cutlass/build/CMakeFiles/CMakeTmp/CMakeLists.txt:
[cmake]   CUDA_ARCHITECTURES is empty for target "cmTC_89c35".
[cmake] 
[cmake] 
[cmake] CMake Error at libtorch/share/cmake/Caffe2/Modules_CUDA_fix/upstream/FindCUDA/select_compute_arch.cmake:137 (try_run):
[cmake]   Failed to generate test project build system.
[cmake] Call Stack (most recent call first):
[cmake]   libtorch/share/cmake/Caffe2/Modules_CUDA_fix/upstream/FindCUDA/select_compute_arch.cmake:197 (CUDA_DETECT_INSTALLED_GPUS)
[cmake]   libtorch/share/cmake/Caffe2/public/utils.cmake:401 (cuda_select_nvcc_arch_flags)
[cmake]   libtorch/share/cmake/Caffe2/public/cuda.cmake:437 (torch_cuda_get_nvcc_gencode_flag)
[cmake]   libtorch/share/cmake/Caffe2/Caffe2Config.cmake:88 (include)
[cmake]   libtorch/share/cmake/Torch/TorchConfig.cmake:68 (find_package)
[cmake]   examples/00_basic_gemm/CMakeLists.txt:38 (find_package)
[cmake] 

```

As you can see, the special FindCuda from libtorch is called…

The architectures to support can be set manually or automatically. By default auto mode is used.

`CUDA_DETECT_INSTALLED_GPUS` cmake function (from libtorch override of `FindCuda`) is used to detect the architecture in auto mode. It will generate a little exec called `detect_cuda_compute_capabilities`
and run it. The exec is a very small piece of code to query GPU and prints its capacity. Manual compilation and run of the exec show it outputs "8.6" which is expected (GPU being a 3090 RTX). However, I don’t know why, in the context of cmake it crashes (log don’t say why). It should not be an issue as there is a backup code in case the compilation doesn’t work but for some reason it completely crashes.

The easiest workaround to use is to set an env variable `TORCH_CUDA_ARCH_LIST` and avoid the GPU detection, for instance this line works:

```shell
TORCH_CUDA_ARCH_LIST=8.6 cmake .. 
# or more inclusive
# TORCH_CUDA_ARCH_LIST="3.5;5.0;5.2;6.0;6.1;7.0;7.5;8.0;8.6;8.6+PTX" cmake .. 
```

Last modification to do: change C++ version from 11 to 14 as it’s the minimum supported by libtorch.

And then, in this case, you can call libtorch from cutlass examples.

**TL;DR**: 3 modifications are needed to make it run, change the C++ version to 14, declare through env variable the supported gpu architectures and modify a single line in a cmake libtorch script to correctly set the CUDA version to use (at least in the right format).

**Comment by hwu36** [comment]
Thank you for your investigation.

Just a brief summary about cuda versions.

Major version: things can be broken between big version.

Minor version: nothing is going to be broken.  new features can be introduced, performance can be improved.  e.g..  check https://github.com/NVIDIA/cutlass/discussions/495

we have update releases between two minor versions.  they just fix bugs.  usually, no new features, no performance improvement.

**Comment by pommedeterresautee** [comment]
Thank you @hwu36, do you think that the GPU architecture torchlib auto detect issue I experience may be related to the CUDA version I am using (11.7)? I don't know how to dig as the log show nothing meaningful I just know it crashes when it tries to run the little app, and the little app works when run manually. Folder `CMakeTmp` is empty.

Last trace (expanded):
```log
/home/geantvert/workspace/cutlass/libtorch/share/cmake/Caffe2/Modules_CUDA_fix/upstream/FindCUDA/select_compute_arch.cmake(136):  if(CMAKE_CUDA_COMPILER_LOADED )
/home/geantvert/workspace/cutlass/libtorch/share/cmake/Caffe2/Modules_CUDA_fix/upstream/FindCUDA/select_compute_arch.cmake(137):  try_run(run_result compile_result /home/geantvert/workspace/cutlass/build /home/geantvert/workspace/cutlass/build/detect_cuda_compute_capabilities.cu RUN_OUTPUT_VARIABLE compute_capabilities )
```

Regarding the `libtorch` fix above, there is a way to avoid modifying the source code.
It appears that the broken `CUDA_VERSION` variable used to init `CUDA_VERSION_STRING` is a cached variable, meaning it can be overriden through the command line.
The following command works on unmodified libtorch:

```shell
# you may want to provide CUTLASS_NVCC_ARCHS too to sync everything
TORCH_CUDA_ARCH_LIST=8.6  cmake .. -D CUDA_VERSION=11.7
```

**Comment by hwu36** [comment]
> do you think that the GPU architecture torchlib auto detect issue I experience may be related to the CUDA version I am using (11.7, also tried 11.6, same result)? 

I don't think so.

Generally, when building gpu kernels, only select the arch that you are going to run on.  Otherwise, it will bloat up the compilation time, the binary size, etc.

**Comment by pommedeterresautee** [comment]
Tks for the info @hwu36.

Closing the issue as it's solved.

For documentation purpose some torch+cutlass code below inspired from example 00, probably not perfect but works (and yes I have seen the profiler, will investigate it soon ;-) ).

> FWIW, on 3090 RTX, keeping everything default, see speedup of 12% with cutlass VS libtorch `mm` (same for `matmul`)

` basic_gemm.cu`

```cpp
#include <iostream>
#include "helper.h"
#include "cutlass/gemm/device/gemm.h"
#include <torch/torch.h>
#include <chrono>

using precision = float;
using LRowMajor = cutlass::layout::RowMajor;
using Gemm = cutlass::gemm::device::Gemm<precision, LRowMajor, precision, LRowMajor, precision, LRowMajor, precision>;


/// Define a CUTLASS GEMM template and launch a GEMM kernel.
cudaError_t CutlassSgemmNN(
        torch::Tensor &A,
        torch::Tensor &B,
        torch::Tensor &C,
        float alpha,
        float beta) {


    //single-precision CUTLASS GEMM with row-major
    // input matrices and 128x128x8 thread block tile size (chosen by default).

    auto a = A.packed_accessor64<float, 2, torch::RestrictPtrTraits>();
    auto b = B.packed_accessor64<float, 2, torch::RestrictPtrTraits>();
    auto c = C.packed_accessor64<float, 2, torch::RestrictPtrTraits>();
    typename Gemm::TensorRefA ref_A;
    typename Gemm::TensorRefB ref_B;
    typename Gemm::TensorRefC ref_C;
    typename Gemm::TensorRefD ref_D;

    // stride is used to extract the leading dimension of each matrix, in our case it's always the nb of cols
    // as we are row oriented.
    ref_A = typename Gemm::TensorRefA((float *) A.data_ptr(), LRowMajor(a.stride(0)));
    ref_B = typename Gemm::TensorRefB((float *) B.data_ptr(), LRowMajor(b.stride(0)));
    ref_C = typename Gemm::TensorRefC((float *) C.data_ptr(), LRowMajor(c.stride(0)));
    ref_D = typename Gemm::TensorRefD((float *) C.data_ptr(), LRowMajor(c.stride(0)));

    int M = (int) A.size(0);
    int K = (int) A.size(1);
    int N = (int) B.size(1);

    // D = A * B + C
    typename Gemm::Arguments arguments{
            {M, N, K},  // Gemm Problem dimensions
            ref_A, // Tensor-ref for source matrix A
            ref_B, // ... B
            ref_C, // ... C
            ref_D, // ... D (may be different memory than source C matrix, not the case here)
            {alpha, beta}};  // Scalars used in the Epilogue

    Gemm gemm_operator;
    cutlass::Status status = gemm_operator(arguments);

    return status == cutlass::Status::kSuccess ? cudaSuccess : cudaErrorUnknown;
}

cudaError_t TestCutlassGemm(int M, int N, int K, float alpha, float beta, double atol = 1e-2) {
    cudaError_t result;

    auto options = torch::TensorOptions()
            .dtype(torch::kFloat32)
            .layout(torch::kStrided)
            .device(torch::kCUDA, 0)
            .requires_grad(false);

    auto tensor_a = torch::rand({M, K}, options);
    auto tensor_b = torch::rand({K, N}, options);
    auto tensor_c_cutlass = torch::empty({M, N}, options);
    auto tensor_c_torch = torch::empty({M, N}, options);


    clock_t start, end;
    int nb_repeat = 10;
    for(int i=0; i<nb_repeat; i++){
        result = CutlassSgemmNN(tensor_a, tensor_b, tensor_c_cutlass, alpha, beta);
    }
    cudaDeviceSynchronize();
    start = clock();
    for(int i=0; i<nb_repeat; i++){
        result = CutlassSgemmNN(tensor_a, tensor_b, tensor_c_cutlass, alpha, beta);
    }
    cudaDeviceSynchronize();
    end = clock();
    auto cutlass_time = (double)(end - start) / CLOCKS_PER_SEC;
    std::cout << "cutlass time: " << cutlass_time << std::endl;

    for(int i=0; i<nb_repeat; i++){
        torch::mm_out(tensor_c_torch, tensor_a, tensor_b);
    }
    torch::cuda::synchronize();
    start = clock();
    for(int i=0; i<nb_repeat; i++){
        torch::mm_out(tensor_c_torch, tensor_a, tensor_b);
    }
    torch::cuda::synchronize();
    end = clock();
    auto torch_time = (double)(end - start) / CLOCKS_PER_SEC;
    std::cout << "torch time: " << torch_time << std::endl;
    std::cout << "speedup: " << torch_time / cutlass_time << std::endl;

    if (result != cudaSuccess) {
        std::cerr << "CUTLASS GEMM kernel failed: " << cudaGetErrorString(result) << std::endl;

        cudaFree(tensor_a.data_ptr());
        cudaFree(tensor_b.data_ptr());
        cudaFree(tensor_c_cutlass.data_ptr());
        cudaFree(tensor_c_torch.data_ptr());

        return result;
    }

    std::cout << std::boolalpha;
    std::cout << "distance torch/cutlass < "
              << atol
              << ": "
              << torch::allclose(tensor_c_torch, tensor_c_cutlass, 0, atol)
              << std::endl;

    cudaFree(tensor_a.data_ptr());
    cudaFree(tensor_b.data_ptr());
    cudaFree(tensor_c_cutlass.data_ptr());
    cudaFree(tensor_c_torch.data_ptr());

    return cudaSuccess;
}

///////////////////////////////////////////////////////////////////////////////////////////////////
// usage:
//
//   00_basic_gemm <M> <N> <K>
//
int main(int argc, const char *arg[]) {
    torch::manual_seed(123);
    // default problem dimensions
    int problem[3] = {32, 512, 1024};
    for (int i = 1; i < argc && i < 4; ++i) {
        std::stringstream ss(arg[i]);
        ss >> problem[i - 1];
    }

    for (int i = 4; i < argc && i < 6; ++i) {
        std::stringstream ss(arg[i]);
    }

    cudaError_t result = TestCutlassGemm(
            problem[0],
            problem[1],
            problem[2],
            1, // alpha
            0 // beta
    );

    return result == cudaSuccess ? 0 : -1;
}
```

And its `CMakeLists.txt`

```cmake
set(CMAKE_PREFIX_PATH "/.../cutlass/libtorch")
set(Torch_DIR "/.../cutlass/libtorch")
find_package(Torch REQUIRED)

cutlass_example_add_executable(
  00_basic_gemm
  basic_gemm.cu
)

target_link_libraries(00_basic_gemm PRIVATE "${TORCH_LIBRARIES}")
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${TORCH_CXX_FLAGS}")
```
